zcash_client_sqlite/src/wallet/transparent/ephemeral.rs

```diff
@@ -0,0 +1,356 @@
+    // its use is at present unmined. This should never occur in supported
+    // situations where only a single correctly operating wallet instance is
+    // using a given seed, because such a wallet will not reuse an address that
+    // it ever reserved.
+    //
+    // `COALESCE(used_in_tx, mined_in_tx)` can only differ from `used_in_tx`
+    // if the address was reserved, an error occurred in transaction creation
+    // before calling `mark_ephemeral_address_as_used`, and then we observed
+    // the address to have been used in a mined transaction (presumably by
+    // another wallet instance, or due to a bug) anyway.
+    let res = wdb
+        .conn
+        .0
+        .query_row(
+            "SELECT t.txid FROM ephemeral_addresses
+             LEFT OUTER JOIN transactions t
+             ON t.id_tx = COALESCE(used_in_tx, mined_in_tx)
+             WHERE address = :address",
+            named_params![":address": address_str],
+            |row| row.get::<_, Option<Vec<u8>>>(0),
+        )
+        .optional()?;
+
+    match res {
+        Some(Some(txid_bytes)) => {
+            let txid = TxId::from_bytes(
+                txid_bytes
+                    .try_into()
+                    .map_err(|_| SqliteClientError::CorruptedData("invalid txid".to_owned()))?,
+            );
+            Err(SqliteClientError::EphemeralAddressReuse(
+                address_str.to_owned(),
+                Some(txid),
+            ))
+        }
+        Some(None) if reject_all_ephemeral => Err(SqliteClientError::EphemeralAddressReuse(
+            address_str.to_owned(),
+            None,
+        )),
+        _ => Ok(()),
+    }
+}
+
+/// If `address` is one of our ephemeral addresses, mark it as having an output
+/// in a transaction that we have just created. This has no effect if `address` is
+/// not one of our ephemeral addresses.
+///
+/// Returns a `SqliteClientError::EphemeralAddressReuse` error if the address was
+/// already used.
+pub(crate) fn mark_ephemeral_address_as_used<P: consensus::Parameters>(
+    wdb: &mut WalletDb<SqlTransaction<'_>, P>,
+    ephemeral_address: &TransparentAddress,
+    tx_ref: i64,
+) -> Result<(), SqliteClientError> {
+    let address_str = encode_transparent_address_p(&wdb.params, ephemeral_address);
+    ephemeral_address_check_internal(wdb, &address_str, false)?;
+
+    wdb.conn.0.execute(
+        "UPDATE ephemeral_addresses SET used_in_tx = :used_in_tx WHERE address = :address",
+        named_params![":used_in_tx": &tx_ref, ":address": address_str],
+    )?;
+    Ok(())
+}
+
+/// If `address` is one of our ephemeral addresses, mark it as having an output
+/// in the given mined transaction (which may or may not be a transaction we sent).
+///
+/// `tx_ref` must be a valid transaction reference. This call has no effect if
+/// `address` is not one of our ephemeral addresses.
+pub(crate) fn mark_ephemeral_address_as_mined<P: consensus::Parameters>(
```

Comment:
The check warnings that say that this method is not covered by tests make me nervous.

Reply:
If this warning is accurate, it means that we have no tests that call `store_decrypted_tx` for *any* transaction that has transparent outputs, when "transparent-inputs" is enabled.

Reply:
I believe the test for the issue @str4d points out in https://github.com/zcash/librustzcash/pull/1257/files#r1661663472 will necessarily exercise this, though.